components/tikv_util/src/logger/file_log.rs

```diff
@@ -230,12 +193,118 @@ impl Rotator for RotateBySize {
     }
 }
 
+#[derive(Debug)]
+pub enum Task {
+    Archive,
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+struct LogInfo {
+    f: DirEntry,
+    dt: DateTime<Local>,
+}
+
+fn dt_from_file_name(path: &Path, prefix: &str) -> Option<DateTime<Local>> {
+    let file_name_with_dt = path.file_stem().unwrap();
+    let mut dt = file_name_with_dt
+        .to_str()
+        .unwrap()
+        .to_string()
+        .replace(prefix, "");
+    if let Some(offset) = dt.find('T') {
+        unsafe {
+            // -2021-12-13T16-08-27.621 => -2021-12-13T16:08:27.621
+            let dt = dt.as_bytes_mut();
+            if dt.len() < offset + 6 {
+                return None;
+            }
+            dt[offset + 3] = b':';
+            dt[offset + 6] = b':';
+        };
+        dt.push_str("+00:00");
+        match DateTime::parse_from_rfc3339(&dt.as_str()[1..]) {
+            Ok(t) => return Some(t.with_timezone(&Local)),
+            Err(_) => return None,
+        }
+    }
+    None
+}
+
+pub struct Runner {
+    log_dir: PathBuf,
+    file_name: String,
+    max_backups: usize,
+    max_days: ReadableDuration,
+}
+
+impl Runner {
+    pub fn new<P: AsRef<Path>>(path: P, max_backups: usize, max_days: ReadableDuration) -> Runner {
+        let mut path = path.as_ref().to_owned();
+        let file_name = path.file_stem().unwrap().to_str().unwrap().to_owned();
+        assert!(path.pop());
+        Runner {
+            log_dir: path,
+            file_name,
+            max_backups,
+            max_days,
+        }
+    }
+
+    fn list_old_logs(&self) -> Result<Vec<LogInfo>, Error> {
+        let mut logs = Vec::new();
+        for f in fs::read_dir(&self.log_dir)? {
+            let f = f?;
+            if f.file_type()?.is_file() {
+                if let Some(dt) = dt_from_file_name(f.path().as_path(), &self.file_name) {
+                    logs.push(LogInfo { f, dt });
+                }
+            }
+        }
+        logs.sort_by(|l1, l2| l2.dt.cmp(&l1.dt));
+        Ok(logs)
+    }
+}
+
+impl Runnable for Runner {
+    type Task = Task;
+
+    fn run(&mut self, _: Task) {
+        if self.max_backups == 0 && self.max_days.is_zero() {
+            return;
+        }
+        let mut logs = self.list_old_logs().unwrap();
+        let mut remove = Vec::new();
+        if self.max_backups > 0 && self.max_backups < logs.len() {
+            remove = logs.split_off(self.max_backups);
+        }
+        if !self.max_days.is_zero() {
+            let threshold = Local::now() - Duration::from_std(self.max_days.0).unwrap();
+            while let Some(log) = logs.last() {
+                if log.dt < threshold {
+                    remove.push(logs.pop().unwrap());
+                } else {
+                    break;
+                }
+            }
+        }
+        for log in remove {
+            fs::remove_file(log.f.path()).unwrap();
```

Comment:
I think we don't need to panic here. Instead, we can log the error.

Reply:
Sure, fixed, thanks!